### ISSUE TITLE
Nanoc is stuck with DDMetrics::Stopwatch::AlreadyRunningError

### ISSUE THREAD

**Issue by axilleas** [nanoc/nanoc#1445]
### Steps to reproduce

I haven't tried to reproduce locally, this has happened while building Nanoc in GitLab CI, and this isn't something that happens consistently.

- Downstream issue https://gitlab.com/gitlab-org/gitlab-docs/issues/395.
- CI log https://gitlab.com/gitlab-org/gitlab-docs/-/jobs/246070460

### Expected behavior

Nanoc builds the site successfully.

### Actual behavior

Nanoc fails to build the site.

### Details

Sometimes, Nanoc is choking on ddmetrics and it errors out. The process doesn't stop,  it rather is stuck and then the CI job fails due to timeout https://gitlab.com/gitlab-org/gitlab-ee/-/jobs/264398546.

Sometimes, it also creates some skip entries https://gitlab.com/gitlab-org/gitlab-docs/-/jobs/246070460.

### Crash log

```
#<Thread:0x0000559eddf3fb20@/builds/gitlab-com/gitlab-docs/vendor/ruby/2.6.0/gems/nanoc-core-4.11.5/lib/nanoc/core/notification_center.rb:29 run> terminated with exception (report_on_exception is true):
/builds/gitlab-com/gitlab-docs/vendor/ruby/2.6.0/gems/ddmetrics-1.0.1/lib/ddmetrics/stopwatch.rb:29:in `start': Cannot start, because stopwatch is already running (DDMetrics::Stopwatch::AlreadyRunningError)
	from /builds/gitlab-com/gitlab-docs/vendor/ruby/2.6.0/gems/nanoc-4.11.5/lib/nanoc/cli/commands/compile_listeners/file_action_printer.rb:32:in `block in start'
	from /builds/gitlab-com/gitlab-docs/vendor/ruby/2.6.0/gems/nanoc-4.11.5/lib/nanoc/cli/commands/compile_listeners/abstract.rb:55:in `block in on'
	from /builds/gitlab-com/gitlab-docs/vendor/ruby/2.6.0/gems/nanoc-core-4.11.5/lib/nanoc/core/notification_center.rb:40:in `block (3 levels) in start'
	from /builds/gitlab-com/gitlab-docs/vendor/ruby/2.6.0/gems/nanoc-core-4.11.5/lib/nanoc/core/notification_center.rb:39:in `each'
	from /builds/gitlab-com/gitlab-docs/vendor/ruby/2.6.0/gems/nanoc-core-4.11.5/lib/nanoc/core/notification_center.rb:39:in `block (2 levels) in start'
	from /builds/gitlab-com/gitlab-docs/vendor/ruby/2.6.0/gems/nanoc-core-4.11.5/lib/nanoc/core/notification_center.rb:32:in `loop'
	from /builds/gitlab-com/gitlab-docs/vendor/ruby/2.6.0/gems/nanoc-core-4.11.5/lib/nanoc/core/notification_center.rb:32:in `block in start'
```

**Comment by ddfreyne** [comment]
This one seems particularly hard to debug. I’m short on time, but will try to debug this more thoroughly soon.

**Comment by axilleas** [comment]
Yeah, I don't have a way to reproduce this unfortunately :/ Thanks, and no rush :)

**Comment by ddfreyne** [comment]
> both packages have the same author

Yep, me :)

> it looks like notification_center is meant to be a single background thread. So I assume it's all fine.

It should be — and since all notifications are handled in order (a `Queue` is FIFO), everything is handled on a single thread in order, so the lack of thread-safety of DDMetrics shouldn’t be an issue in this case.

However, there *is* some threading going on in the writer (`Nanoc::Int::Compiler::Phases::Write`), which has a threadpool to make more efficient use of I/O. The `ItemRepWriter` sends the `rep_write_started` and `rep_write_ended` notifications, which are sent from a different thread, which means that the problem could be that the notification center is not thread-safe. It uses a `Queue`, though, which is threadsafe.

This specific crash happens on `file_action_printer.rb` line `32`, which is `@stopwatches[rep].start`. The crash happens when this particular stopwatch is already started, meaning that the call to this particular `#start` (caused by a `:rep_write_started` notification) is expected to be preceded by a `:rep_write_started` notification, but is not. Possible culprit is here:

```
# Nanoc::Int::Compiler::Phases::Write
# write.rb lines 75-7

            @queue << rep

            Nanoc::Core::NotificationCenter.post(:rep_write_enqueued, rep)
```

I think these two lines need to be swapped: notify *before* enqueuing.

I unfortunately have no way of testing it out properly, and it’s a bit late (brain is shutting off), but if either of you want to give this a try and see whether swapping those two lines around, that’d be of immense help.

Will continue investigating later!

**Comment by ddfreyne** [comment]
I’ve pushed the aforementioned change to `master`, which hopefully fixes the problem. Can you try out nanoc from `master` and see whether it works?

* * *

To do so, edit `Gemfile` and change the `nanoc` line to say

```
gem 'nanoc', github: 'nanoc/nanoc`
````

Then run `bundle update nanoc`, and commit the changes to `Gemfile` and `Gemfile.lock`. CI should now be using Nanoc from GitHub.

**Comment by axilleas** [comment]
> So although I don't have an easily reproducible case, perhaps trying to build the gitlab docs on a wimpy machine or heavily niced would suffice?

The error has also happened on a beefy machine, but I'll keep that in mind.

> I’ve pushed the aforementioned change to master, which hopefully fixes the problem. Can you try out nanoc from master and see whether it works?

Thanks! We'll try that (https://gitlab.com/gitlab-org/gitlab-docs/merge_requests/531) and let you know :)

**Comment by sirosen** [comment]
I've been trying to reproduce the issue with the change in place, but so far haven't been able to do so. I hesitate to promise that it's gone for good since it was not 100% reproducible in the first place, but I'll continue to poke at it and see if I can trigger it again.

**Comment by axilleas** [comment]
Likewise, we're running this in production with no errors so far. I'll continue to keep an eye on it :)

**Comment by sirosen** [comment]
Through the night here, I let my own machine run two hundred parallel builds here, half running the master branch, half running v4.11.9 .
Literally just `(set -e; for i in {0..100}; do echo "$i" > counter.txt; make rebuild; done)` in two separate copies of our site repo.

When I came back to check it this morning...
```
$ cat counter.txt
1
```
so it built twice and failed on number 3.

In the version running master
```
$ cat counter.txt
100
```

I can more trials of this experiment, but even just doing that once leaves me quite comfortable saying that this was the bug and it's fixed now.

**Comment by ddfreyne** [comment]
@sirosen Thank you for the thorough testing! I’ll get a patch release out this weekend.

**Comment by ddfreyne** [comment]
This is released in Nanoc 4.11.10.
